oeps/oep-0017-feature-toggles.rst

```diff
@@ -0,0 +1,809 @@
++-----------------+--------------------------------------------------+
+| Type            | Best Practice                                    |
++-----------------+--------------------------------------------------+
+| Created         | 2018-02-14                                       |
++-----------------+--------------------------------------------------+
+| `Review Period` | 2018-02-14 - 2018-03-07                          |
++-----------------+--------------------------------------------------+
+| `Resolution`    |                                                  |
++-----------------+--------------------------------------------------+
+
+.. contents::
+   :local:
+   :depth: 2
+
+Abstract
+========
+
+A feature toggle is a software development technique that decouples 
+deployment of code from release (enablement) of the code. When used with care, 
+it can be a powerful tool in a continuous deployment environment to deploy 
+changes incrementally with the main codebase while under development. 
+Additionally, it allows a team to reduce the risk of breaking 
+production systems by providing levers to progressively test changes and quickly 
+disable changes if needed. For a large platform, it is also used to 
+selectively enable a change/feature for certain users or certain deployments.
+
+Feature toggles inherently add complexity to the code, however, since they 
+introduce multiple paths and configuration operations, with the potential to
+create an explosion in testing permutations and to create forgotten latent unused
+code paths. A thorough understanding of their implications and best practices is
+warranted to use them in a large-scale system.
+
+Motivation
+==========
+
+Given the many benefits of using feature toggles, edX development teams have
+been using them since the very start. Additionally, since deployment latency
+for edx.org has reduced from monthly->weekly->daily and since teams have been
+bitten too many times by `long-term feature branches`_, teams now have a greater 
+incentive to develop incrementally within the main branch. Updating veteran 
+edX user experiences and features also drives the need for a framework where 
+usability and behavior changes are gradually released and introduced to the edX 
+user base.
+
+Aligning on a common best practice for feature toggling is critical for both
+the short-term and long-term health of the system. Using different standards,
+strategies, testing procedures, etc leads to confusion, production failures,
+and long-term maintenance issues.
+
+.. _long-term feature branches: https://blog.newrelic.com/2012/11/14/long-running-branches-considered-harmful/
+
+Use Cases
+---------
+
+At its core, feature toggles allow teams to deploy alternative code paths and to
+choose between them at runtime. There are multiple scenarios where this capability
+comes in handy.
+
+**Note:** Using feature toggles for experimentation is out of the scope for this
```

Comment:
We did use it and I can imagine using it again.  Not sure if you want it to be a separate use case, or it could be an exception to the "Incremental Release" use case where we are more ok with it because we plan to quickly remove it?  Or, do you think something else is what we should have done?